### DISCUSSION TITLE
Async method passing `self` using `belonging_to`

### DISCUSSION THREAD

**Discussion by perseoGI** [diesel-rs/diesel#3649] [Q&A] [answered]
I there, I've been strugling with diesel v2 and async code.

Imagine following code:

```rust
#[derive(Debug, Identifiable, Queryable, Associations, Deserialize, Serialize)]
#[diesel(table_name = images)]
#[diesel(belongs_to(Part, foreign_key = part_id))]
pub struct Image {
    pub id: Id,
    pub part_id: Id,
    pub path: String,
}

#[derive(Identifiable, Queryable, Serialize)]
pub struct Part {
    pub id: Id,
    ...
}
#[derive(Queryable, Debug, Serialize)]
pub struct PartDTO {
    pub id: Id,
    pub images: Vec<(Id, String)>,
}
```

Lets say I want to use the `belonging_to` method to query all images belonging to a part this way:

```rust
pub async fn to_dto(&self, conn: &DbConn) -> PartDTO {
    let images: Vec<(Id, String)> = conn.run(move |c| {
        Image::belonging_to(self)
            .get_results::<Image>(c)
    })
    .await
    .unwrap()
    .into_iter()
    .map(|image| {
        (
            image.id,
            Path::new(&image.path)
                .file_stem()
                .unwrap()
                .to_os_string()
                .to_str()
                .unwrap()
                .into(),
        )
    })
    .collect();

  PartDTO {id: self.id, images}
} 
```

So that, in any services (such as create, get_one, get_all, etc...) I could end up calling thes `to_dto` method, eg of create:
```rust
pub async fn create(part: PartInputDTO, conn: &DbConn) -> QueryResult<PartDTO> {
       let created_part = conn
        .run(move |c| {
            diesel::insert_into(parts)
                .values(&part_input)
                .get_result::<Part>(c)
        })
        .await;

    Ok(created_part.unwrap().to_dto(conn).await)
}
```

Sadly, I get the following compile time error:

```
self` is a reference that is only valid in the method body let's call the lifetime of this reference `'1`
`self` escapes the method body here (conn.run)  argument requires that `'1` must outlive `'static`
```
I've tried making the `self` parameter `&'static Self`, using `std::rc::Rc pointers` to clone the self object, using `async-trait` lib, but all of those solutions seems to fail. 

I guess I could avoid using `belonging_to` function and use a standard query, but this code used to work in diesel v1 and I would love to keep it using and understand how to fix this.

Thanks in advance!

`diesel version == 2.1.0`

**Comment by weiznich** [comment]
That's only semi-diesel related, as this issue is caused by the `run` function which is not provided by diesel. In short the problem there is that this function spawn work on a possibly different thread, which means all captures of a given closure need to fulfill the `'static` lifetime bound. This is nothing that's reasonable possible with that kind of setup.

**Reply by perseoGI** [reply, marked as answer]
Thank you!

As explained in [here](https://github.com/SergioBenitez/Rocket/discussions/2556), the only reasonable way of achieving this is by cloning current `part` instance and using it to query the DB (with `belonging_to`).

For that, both `Part` and `Image` structures must derive from `copy`
